components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt

```diff
@@ -0,0 +1,90 @@
+package mozilla.components.feature.downloads.manager
+
+import android.Manifest.permission.FOREGROUND_SERVICE
+import android.Manifest.permission.INTERNET
+import android.Manifest.permission.WRITE_EXTERNAL_STORAGE
+import android.app.DownloadManager.ACTION_DOWNLOAD_COMPLETE
+import android.app.DownloadManager.EXTRA_DOWNLOAD_ID
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import android.util.LongSparseArray
+import androidx.annotation.RequiresPermission
+import androidx.core.util.isEmpty
+import androidx.core.util.set
+import androidx.localbroadcastmanager.content.LocalBroadcastManager
+import mozilla.components.browser.session.Download
+import mozilla.components.feature.downloads.AbstractFetchDownloadService
+import mozilla.components.feature.downloads.ext.putDownloadExtra
+import kotlin.random.Random
+
+class FetchDownloadManager<T : AbstractFetchDownloadService>(
+    private val applicationContext: Context,
+    private val service: Class<T>,
+    private val broadcastManager: LocalBroadcastManager = LocalBroadcastManager.getInstance(applicationContext),
+    override var onDownloadCompleted: OnDownloadCompleted = noop
+) : BroadcastReceiver(), DownloadManager {
+
+    private val queuedDownloads = LongSparseArray<Download>()
+    private var isSubscribedReceiver = false
+
+    override val permissions = arrayOf(INTERNET, WRITE_EXTERNAL_STORAGE, FOREGROUND_SERVICE)
+
+    @RequiresPermission(allOf = [INTERNET, WRITE_EXTERNAL_STORAGE, FOREGROUND_SERVICE])
+    override fun download(download: Download, cookie: String): Long? {
+        if (!download.isScheme(listOf("http", "https"))) {
```

Comment:
They're some validations that I noticed that are duplicated between the Android download manger and this one. Maybe in the feature we may want to share validations.

Reply:
This validation will change in the future as GeckoView gets support for blob URLs (and possibly other schemes). I tried to share most of the validation code and allow for a custom scheme list.